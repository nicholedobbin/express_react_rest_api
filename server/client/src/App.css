
/* App div: Align and style font. */
.App {
  text-align: center;
  font-family: 'Helvetica Neue', Arial, Helvetica, sans-serif;
  background-color: #eee;
}

/* Header Section: set colours and style fonts. */
.headerSection {
  background-color: #4c5357;
  color: #fff;
  font-family: 'Helvetica Neue', Arial, Helvetica, sans-serif;
}

.headerSection p{
  font-size: 18px;
}

/* Search button: set colours */
.searchButton {
  background-color:  #5fd3ef;
  border-color: #5fd3ef;
}

/* GitHub/GitLab choice div/tabs: set colours, display, padding, width and alignment. */
.choiceDiv {
  background-color: #7bc74d;
  display: flex;
  padding: 10px;
  border-radius: 0.375rem;
  width: 27%;
  justify-content: center;
}

.choiceItem {
  padding: 15px;
  color: #4c5357;
  font-size: 20px;
  font-weight: 300;
  line-height: 1.2;
}

.choiceItem :hover{
  color: white;
  cursor: pointer;
}

.selected {
  font-weight: bold;
  color: white;
}

/* Results user profile section: set colours and width. */
.resultsCard {
  width: 35rem;
}

.profileName {
  color: #5fd3ef;
}

.profileName :hover{
  color: #7bc74d;
}


/* Results user repos section: set repos and commits background colours, add margins and padding, style font.*/
.repoList {
  background: rgba(95, 211, 239, .1);
  padding: 15px;
  margin: 10px;
  color: #4c5357;
}

.repoCommits {
  background: rgb(123, 199, 77, .1)!important;
}

.repoTitle {
  color: #4c5357;
  font-size: 18px;
  text-align: center;
  font-weight: bold;
  padding-bottom: 20px;
  padding-top: 10px;
}

.reposHeading{
  color: #5fd3ef;
}
