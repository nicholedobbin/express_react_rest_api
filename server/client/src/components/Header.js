import React from 'react'

export default function Header() {
  return (
        <div className="headerSection py-5">
            <h1 className="headerTitle py-3  display-2">GitHub & GitLab API Search</h1>
            <p className="headerDescription w-75 mx-auto py-3">
                Find a user from the GitHub & GitLab API's by entering the username in the search bar.
                Switch between the 'View GitHub' and 'View GitLab' options to view profile and repo details.  
            </p>
        </div>
  )
}
