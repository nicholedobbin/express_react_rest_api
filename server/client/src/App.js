import './App.css';
import { useState } from 'react';

// Component imports
import SearchBar from './components/SearchBar';
import GitHubResults from './components/GitHubResults';
import Header from './components/Header';
import GitLabResults from './components/GitLabResults';

// Boostrap imports
import ListGroup from 'react-bootstrap/ListGroup';
import ListGroupItem from 'react-bootstrap/cjs/ListGroupItem';

function App() {
  // --------------- Set State ---------------
  // GitHub and GitLab user profiles (all data from server).
  const [userProfiles, setUserProfiles] = useState();

  // GitHub profile and repo data.
  const [gitHubAvatar, setGitHubAvatar] = useState();
  const [gitHubName, setGitHubName] = useState();
  const [gitHubProfileLink, setGitHubProfileLink] = useState();
  const [gitHubBio, setGitHubBio] = useState();
  const [gitHubRepoData, setGitHubRepoData] = useState();

  // GitLab profile and repo data.
  const [gitLabAvatar, setGitLabAvatar] = useState();
  const [gitLabName, setGitLabName] = useState();
  const [gitLabProfileLink, setGitLabProfileLink] = useState();
  const [gitLabRepoData, setGitLabRepoData] = useState();

  // General states.
  const [searchName, setSearchName] = useState("");
  const [option, setOption] = useState();
  const [loading, setLoading] = useState(false);
  const [loadingRepos, setLoadingRepos] = useState(false);

  // --------------- Get User Profiles ---------------
  async function getUser() {
    setLoading(true);
    // Get all GitHub and GitLab profiles data for inputted username.
    let response = await fetch(`/searchUser/${searchName}`);
    let profilesData = await response.json();

    // Set userProfiles state to store GitHub and GitLab profiles data.
    setUserProfiles(profilesData.data);
    
    // set GitHub profile details.
    setGitHubName(profilesData.data[0].gitHubUserName);
    setGitHubProfileLink(profilesData.data[0].gitHubProfileLink);
    setGitHubAvatar(profilesData.data[0].gitHubAvatar);
    setGitHubBio(profilesData.data[0].gitHubBio);

    // set GitLab profile details.
    setGitLabName(profilesData.data[1].gitLabUserName);
    setGitLabProfileLink(profilesData.data[1].gitLabProfileLink);
    setGitLabAvatar(profilesData.data[1].gitLabAvatar);

    setLoading(false);
    // Call function to get repos for the inputted username.
    getRepos(searchName);
  }

  // --------------- Get User Repos ---------------
  async function getRepos(searchName) {
    setLoadingRepos(true);
    // Get all GitHub and GitLab repo data for inputted username.
    let response = await fetch(`/searchUser/${searchName}/repos`);
    let repoData = await response.json();

    // create seperate repoLists to store data for GitHub and GitLab repos.
    let gitHubRepoList = repoData.data[0].gitHubProjects;
    let gitLabRepoList = repoData.data[1].gitLabProjects;

    // Loop through gitHubRepoList:
    for (let i = 0; i < gitHubRepoList.length; i++) {
      // If the gitHubRepoList item's repoFound value is true:
      if (gitHubRepoList[i].repoFound === true) {
        // Map through gitHubRepoList and the commitsMessages for that repo and return relevant details.
        let gitHubRepoInfo = gitHubRepoList.map((repo, index) => {
          return (
            <ListGroup key={index} className="text-start">
              <ListGroup.Item className="repoList">
                <div className="repoTitle">{repo.repoName}</div>
                <div className="pb-2">Created on: {repo.dateCreated}</div>
                <div className="repoDescription pb-2">
                  Description:
                  <br></br>
                  {repo.repoDescription}
                </div>
                <div>
                  <span className="latestCommitsHeading">Latest Commit Messages:</span>
                  <br></br>
                  {repo.commitMessages.map((message, index) => {
                    return (
                      <ListGroup.Item className="repoCommits" key={index}>
                        {message.commitDate}:
                        <br></br>
                        {message.commitMsg} 
                      </ListGroup.Item>
                    )
                  })}
                </div>
              </ListGroup.Item>
            </ListGroup>
          )
        })

        // Set gitHubRepoData to value of gitHubRepoInfo.
        setGitHubRepoData(gitHubRepoInfo);

      } else {
        // Else, create noReposMsg React element and set gitHubRepoData to this value. 
        let noReposMsg = (
          <ListGroup>
            <ListGroupItem>No GitHub repos found for this user </ListGroupItem>
          </ListGroup>
        );
        setGitHubRepoData(noReposMsg);
      }
    }

    // Loop through gitLabRepoList:
    for (let i = 0; i < gitLabRepoList.length; i++) {
      // If the gitLabProjects' repoFound value is true:
      if (gitLabRepoList[i].repoFound === true) {
        // Map through gitLabRepoList and the commitsMessages for that repo and return relevant details.
        let gitLabRepoInfo = gitLabRepoList.map((repo, index) => {
          return (
            <ListGroup key={index} className="text-start">
              <ListGroup.Item className="repoList">
                <div className="repoTitle">{repo.repoName}</div>
                <div className="pb-2">Created on: {repo.dateCreated}</div>
                <div className="repoDescription pb-2">
                  Description:
                  <br></br>
                  {repo.repoDescription}
                </div>
                <div>
                  <span className="latestCommitsHeading">Latest Commit Messages:</span>
                  <br></br>
                  {repo.commitMessages.map((message, index) => {
                    return (
                      <ListGroup.Item className="repoCommits" key={index}>
                        {message.commitDate}:
                        <br></br>
                        {message.commitMsg} 
                      </ListGroup.Item>
                    )
                  })}
                </div>
              </ListGroup.Item>
            </ListGroup>
          )
        })

        // Set gitLabRepoData to value of gitLabRepoInfo.
        setGitLabRepoData(gitLabRepoInfo);

      } else {
        // Else, create noReposMsg React element and set gitHubRepoData to this value. 
        let noReposMsg = (
          <ListGroup>
            <ListGroupItem>No GitLab repos found for this user</ListGroupItem>
          </ListGroup>
        );
        setGitLabRepoData(noReposMsg);
      }
    }
    // setLoading(false);
    setLoadingRepos(false);
  }

  // Handle change function for SearchBar component's onChange value.  
  const handleChange = (e) => {
    setSearchName(e.target.value);
  }
  // Handle option function for SearchBar component's onClick value.
  const handleOption = (e) => {
    setOption(e);
  }

  return (
    <div className="App">
      <Header />
      <div className="App justify-content-center align-items-center d-flex flex-column">
        <SearchBar
          getUser = {getUser}
          handleChange = {handleChange}
          searchName = {searchName}
          option = {option}
          handleOption = {handleOption}
        />
        { loading ? <p>Loading...</p> : option ? <GitLabResults
          userProfiles = {userProfiles}
          gitLabAvatar = {gitLabAvatar}
          gitLabName = {gitLabName}
          gitLabProfileLink = {gitLabProfileLink}
          searchName = {searchName}
          gitLabRepoData = {gitLabRepoData}
          loadingRepos = {loadingRepos}
        /> :
          <GitHubResults
            userProfiles = {userProfiles}
            gitHubAvatar = {gitHubAvatar}
            gitHubName = {gitHubName}
            gitHubProfileLink = {gitHubProfileLink}
            gitHubBio = {gitHubBio}
            searchName = {searchName}
            gitHubRepoData = {gitHubRepoData}
            loadingRepos = {loadingRepos}
          /> }
      </div>
    </div>
  );
}

export default App;
