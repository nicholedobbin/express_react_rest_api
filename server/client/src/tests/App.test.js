import React from "react";
import Header from "../components/Header";
import renderer from 'react-test-renderer';
import App from '../App';

const getRepos = require('../App');

// Snapshot test: tests if Header component renders correctly.
test('Header component renders correctly', () => {
  const tree = renderer
  .create(<Header/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Snapshot test: tests if App component renders correctly.
test('main component renders correctly', () => {
  const tree = renderer
.create(<App /> )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Unit test: tests if user is found at specific endpoint.
test('user is found at searchUser endpoint', () => {
  let searchName = process.env.REACT_APP_SEARCH_NAME;
  return fetch(`http://localhost:3000/searchUser/${searchName}`)
  .then(response => response.json())
  .then(data => {
    expect(data.data[0].userFound).toBe(true);
  })
})

// Unit test: tests if repos are found at specific endpoint.
test('projects are found at getRepos endpoint', () => {
  let searchName = process.env.REACT_APP_SEARCH_NAME;
  return fetch(`http://localhost:3000/searchUser/${searchName}/repos`)
  .then(response => response.json())
  .then(data => {
    expect(data.msg).toContain('Repos found for username');
  })
}, 60_000)

// test('projects are found at getRepos endpoint', () => {
//   // jest.setTimeout(60000);
//   let searchName = process.env.REACT_APP_SEARCH_NAME;
//   return fetch(`http://localhost:3000/searchUser/${searchName}/repos`).then(data => {
//     expect(data.data[0].gitHubProjects[0].repoFound).toBe(true);
//   });
// }, 60_000);