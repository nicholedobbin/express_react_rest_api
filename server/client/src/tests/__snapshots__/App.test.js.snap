// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders correctly 1`] = `
<div
  className="headerSection py-5"
>
  <h1
    className="headerTitle py-3  display-2"
  >
    GitHub & GitLab API Search
  </h1>
  <p
    className="headerDescription w-75 mx-auto py-3"
  >
    Find a user from the GitHub & GitLab API's by entering the username in the search bar. Switch between the 'View GitHub' and 'View GitLab' options to view profile and repo details.
  </p>
</div>
`;

exports[`main component renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="headerSection py-5"
  >
    <h1
      className="headerTitle py-3  display-2"
    >
      GitHub & GitLab API Search
    </h1>
    <p
      className="headerDescription w-75 mx-auto py-3"
    >
      Find a user from the GitHub & GitLab API's by entering the username in the search bar. Switch between the 'View GitHub' and 'View GitLab' options to view profile and repo details.
    </p>
  </div>
  <div
    className="App justify-content-center align-items-center d-flex flex-column"
  >
    <div
      className="formSection mt-4 pt-4"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <label
            className="display-6 pb-4 form-label"
            htmlFor="formBasicText"
          >
            Enter username:
          </label>
          <input
            className="mb-4 form-control"
            id="formBasicText"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          className="btn btn-lg searchButton btn btn-secondary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="choiceDiv my-5"
    >
      <div
        className="choiceItem selected"
        onClick={[Function]}
      >
        View GitHub
      </div>
      <div
        className="choiceItem"
        onClick={[Function]}
      >
        View GitLab
      </div>
    </div>
    <p />
  </div>
</div>
`;
